diff --git a/Makefile.in b/Makefile.in
index b301f4d..969db5e 100644
--- a/Makefile.in
+++ b/Makefile.in
@@ -67,9 +67,7 @@
 @ENABLE_FRAME_POINTERS_TRUE@@X86_64_AND_NO_FP_BY_DEFAULT_TRUE@am__append_4 = -fno-omit-frame-pointer
 @ENABLE_FRAME_POINTERS_FALSE@@X86_64_AND_NO_FP_BY_DEFAULT_TRUE@am__append_5 = -DNO_FRAME_POINTER
 @MINGW_TRUE@am__append_6 = -Wl,-u__tcmalloc
-noinst_PROGRAMS = $(am__EXEEXT_1) $(am__EXEEXT_2) $(am__EXEEXT_3) \
-	$(am__EXEEXT_4) $(am__EXEEXT_5) $(am__EXEEXT_6) \
-	$(am__EXEEXT_7) $(am__EXEEXT_8) $(am__EXEEXT_28)
+noinst_PROGRAMS = 
 bin_PROGRAMS =
 @MINGW_TRUE@am__append_7 = libwindows.la libspinlock.la
 
