Only in ./: perftools-stddef.patch
diff -ru .//src/base/vdso_support.h ../google-perftools-1.6-patched/src/base/vdso_support.h
--- .//src/base/vdso_support.h	2010-08-03 03:32:20.000000000 +0800
+++ ../google-perftools-1.6-patched/src/base/vdso_support.h	2011-06-16 14:33:10.000000000 +0800
@@ -38,6 +38,7 @@
 #define HAVE_VDSO_SUPPORT 1
 
 #include <stdlib.h>     // for NULL
+#include <stddef.h>
 #include <link.h>  // for ElfW
 #include "base/basictypes.h"
 
diff -ru .//src/symbolize.h ../google-perftools-1.6-patched/src/symbolize.h
--- .//src/symbolize.h	2009-11-11 03:59:46.000000000 +0800
+++ ../google-perftools-1.6-patched/src/symbolize.h	2011-06-16 14:32:55.000000000 +0800
@@ -38,6 +38,7 @@
 #include <stdint.h>  // for uintptr_t
 #endif
 #include <map>
+#include <stddef.h>
 
 using std::map;
 
diff -ru .//src/system-alloc.h ../google-perftools-1.6-patched/src/system-alloc.h
--- .//src/system-alloc.h	2010-04-01 07:50:18.000000000 +0800
+++ ../google-perftools-1.6-patched/src/system-alloc.h	2011-06-16 14:32:38.000000000 +0800
@@ -38,6 +38,7 @@
 
 #include <config.h>
 #include "internal_logging.h"
+#include <stddef.h>
 
 // REQUIRES: "alignment" is a power of two or "0" to indicate default alignment
 //
