diff --git a/src/profiledata.h b/src/profiledata.h
index 1b3eb2e..0d81603 100644
--- a/src/profiledata.h
+++ b/src/profiledata.h
@@ -106,7 +106,7 @@ class ProfileData {
     int      frequency_;                  // Sample frequency.
   };
 
-  static const int kMaxStackDepth = 64;  // Max stack depth stored in profile
+  static const int kMaxStackDepth = 300;  // Max stack depth stored in profile
 
   ProfileData();
   ~ProfileData();
